repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java

```diff
@@ -1,82 +1,126 @@
+import jakarta.persistence.TypedQuery;
+import jakarta.persistence.criteria.*;
 import org.springframework.stereotype.Repository;
 
 import java.util.List;
 import java.util.Optional;
-
-import static com.epam.esm.util.DatabaseQueryConstants.*;
+import java.util.Set;
 
 /**
- * Repository implementation for handling operations related to Tags in the database.
+ * Implementation of the {@link TagRepository} interface.
  */
-@Slf4j
 @Repository
-@RequiredArgsConstructor
 public class TagRepositoryImpl implements TagRepository {
 
-    private final JdbcTemplate jdbcTemplate;
-    private final TagRowMapper giftTagRowMapper;
+    @PersistenceContext
+    private EntityManager entityManager;
 
     /**
-     * Retrieves all Tags from the database.
-     *
-     * @return List of all Tag objects.
+     * @inheritDoc
      */
     @Override
-    public List<Tag> findAll() {
-        log.info("Querying all rows from gift tag table...");
-        return jdbcTemplate.query(FIND_ALL_TAGS, giftTagRowMapper);
+    public List<Tag> findAllByPage(Pagination pagination) {
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Tag> criteriaQuery = criteriaBuilder.createQuery(Tag.class);
+        Root<Tag> root = criteriaQuery.from(Tag.class);
+
+        criteriaQuery.select(root);
+        TypedQuery<Tag> query = entityManager.createQuery(criteriaQuery);
+        query.setFirstResult(pagination.getOffset());
+        query.setMaxResults(pagination.getLimit());
+
+        return query.getResultList();
     }
 
     /**
-     * Finds a Tag by its ID.
-     *
-     * @param id The ID of the Tag to find.
-     * @return Optional containing the Tag if found, otherwise empty.
+     * @inheritDoc
      */
     @Override
-    public Optional<Tag> findById(Long id) {
-        log.info("Querying a row from gift tag table by id...");
-        return Optional.ofNullable(jdbcTemplate.queryForObject(FIND_TAG_BY_ID, giftTagRowMapper, id));
+    public Tag findById(Long id) {
+        Tag tag = entityManager.find(Tag.class, id);
+        return Optional.ofNullable(tag)
+                .orElseThrow(() -> new EntityNotFoundException("Failed to find tag by id " + id));
+    }
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public Optional<Tag> findByName(String name) {
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Tag> criteriaQuery = criteriaBuilder.createQuery(Tag.class);
+        Root<Tag> root = criteriaQuery.from(Tag.class);
+
+        criteriaQuery.select(root)
+                .where(criteriaBuilder.equal(root.get("name"), name));
+
+        TypedQuery<Tag> query = entityManager.createQuery(criteriaQuery);
+        return query.getResultList().stream().findFirst();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public Tag findMostUsedTagOfUserWithHighestOrderCost(Long userId) {
+        String sql = """
+           SELECT
```

Comment:
In addition, you need to add a query execution plan for this query with a description.

Reply:
@NozimjonNabiev Create new txt file and add query execution plan to it with description